### ISSUE TITLE
rsync cache not preventing duplicate fetches in screwball case

### ISSUE THREAD

**Issue by sraustein** [dragonresearch/rpki.net#108]
Seen in a log today (at log_verbose level):

{{{
$ fgrep NIST screenlog.0  
19:24:06: rcynic: Fetching rsync://rpki.antd.nist.gov/rpki/NIST/1/
19:24:22: rcynic: Fetching rsync://rpki.antd.nist.gov/rpki/NIST/1/
19:24:38: rcynic: Fetching rsync://rpki.antd.nist.gov/rpki/NIST/1/
19:24:54: rcynic: Checking manifest rsync://rpki.antd.nist.gov/rpki/NIST/1/I6OBANx93s5e-gU_FGhY9DJ5UCY.mnf
19:24:54: rcynic: Checking CRL rsync://rpki.antd.nist.gov/rpki/NIST/1/I6OBANx93s5e-gU_FGhY9DJ5UCY.crl
19:24:54: rcynic: Checking ROA rsync://rpki.antd.nist.gov/rpki/NIST/1/fbVTjwEKuN2vI7ykNlivbVtbGgY.roa
19:24:54: rcynic: Checking ROA rsync://rpki.antd.nist.gov/rpki/NIST/1/gR_4W1TqitnDbRzLscPsMLvF6Ks.roa
19:24:54: rcynic: Checking ROA rsync://rpki.antd.nist.gov/rpki/NIST/1/vuo21wXclbO92eRjXYO8coREctI.roa
19:24:54: rcynic: Checking manifest rsync://rpki.antd.nist.gov/rpki/NIST/1/1F3ytMHqtBdR-XE46OaFnhpAJso.mnf
19:24:54: rcynic: Checking CRL rsync://rpki.antd.nist.gov/rpki/NIST/1/1F3ytMHqtBdR-XE46OaFnhpAJso.crl
19:24:55: rcynic: Checking manifest rsync://rpki.antd.nist.gov/rpki/NIST/1/aIVz3x9ZNv1VCo6JNRuPMMClU2U.mnf
19:24:55: rcynic: Checking CRL rsync://rpki.antd.nist.gov/rpki/NIST/1/aIVz3x9ZNv1VCo6JNRuPMMClU2U.crl
19:24:55: rcynic: Fetching rsync://rpki.antd.nist.gov/rpki/NIST/2/
19:24:56: rcynic: Checking manifest rsync://rpki.antd.nist.gov/rpki/NIST/1/Ps5LfCRZAa3itnHvsmMdskw0VlM.mnf
19:24:56: rcynic: Checking CRL rsync://rpki.antd.nist.gov/rpki/NIST/1/Ps5LfCRZAa3itnHvsmMdskw0VlM.crl
19:24:56: rcynic: Checking ROA rsync://rpki.antd.nist.gov/rpki/NIST/1/ILCPpcLFpdPmUMiq0rcRv7S4uHs.roa
19:24:56: rcynic: Checking ROA rsync://rpki.antd.nist.gov/rpki/NIST/1/ZZI0w9ZJVG4sl1y_5bf_Z1NTPx8.roa
19:24:56: rcynic: Checking ROA rsync://rpki.antd.nist.gov/rpki/NIST/1/tgBLaIAJ4j5SuG0DWf_udMnecaw.roa
19:25:11: rcynic: Checking manifest rsync://rpki.antd.nist.gov/rpki/NIST/2/Z_APTY7velsXgE1bqG0TnzUgTcs.mnf
19:25:11: rcynic: Checking CRL rsync://rpki.antd.nist.gov/rpki/NIST/2/Z_APTY7velsXgE1bqG0TnzUgTcs.crl
19:25:11: rcynic: Checking ROA rsync://rpki.antd.nist.gov/rpki/NIST/2/QB1Rm3sa1TaDPcuHWQmidY8WByw.roa
19:25:11: rcynic: Checking ROA rsync://rpki.antd.nist.gov/rpki/NIST/2/ifNYQsUTjGJrUvKbo2tWdQqw0jc.roa
19:25:11: rcynic: Checking ROA rsync://rpki.antd.nist.gov/rpki/NIST/2/sMEgudTnpcQP4bqrOvGNe-e8d0U.roa
}}}

Note that:
1. The three fetches are identical.  This shouldn't happen.  As far as I have seen it doesn't happen for anybody else, and the rsync cache is old code that's been stable for a long time.  Why is NIST special?
2. None of the URLs being checked are duplicates.

My guess is that NIST is special because NIST did something odd when setting up: they repeatedly set up the software, got it as far as receiving RPKI certs issued by its parent, then destroying their installation including all of their keys and doing the whole thing again.  Given the current naming scheme, this would result in multiple issued CAs all pointing at the same directory on NIST's server, which is what it looks like we're seeing.

If this is correct, what's happening here is that we're doing a normal walk of pseudo-ARIN, and hit three separate NIST certs pointing at the same directory.  The conflict detection code keeps us from starting rsync on more than one of them at a time, but the unblocking code doesn't check the rsync cache to see whether the scheduled fetch is now unnecessary.

If this theory is correct, the fix should be straighforward, although there may be some tedious data structure manipulation. Not worth doing just for NIST, but there are other theoretical scenarios where something similar to this could occur, so we should be checking the rsync cache before starting a deferred rsync.

_Trac ticket #98 component rcynic priority minor, owner sra, created by sra on 2011-10-06T19:37:08Z, last modified 2011-10-06T23:41:00Z_
